### PR TITLE
Simple mdns implementation using the Espressif code

### PR DESCRIPTION
**Pull request by pjsg**
This PR is a thin layer over the espressif provided mdns implementation. It only allows a single service to be advertised and it doesn't allow the hostname and the instance name to be broken apart. However, it works, is simple, and reasonably compact.

### CODE REVIEW

File: app/include/user_modules.h [hunk 1]
```diff
@@ -27,6 +27,7 @@
 #define LUA_USE_MODULES_HTTP
 //#define LUA_USE_MODULES_HX711
 #define LUA_USE_MODULES_I2C
+#define LUA_USE_MODULES_MDNS
```
Comment (marcelstoer) [inline app/include/user_modules.h]: I'd rather have this disabled by default even though #872 is still waiting for a resolution.

File: app/modules/mdns.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  return 0;
+}
+
+static struct wrapper *process_args(lua_State *L, int phase, size_t *sizep)
+{
+  struct wrapper *result = NULL;
+  char *p = NULL;
+
+  if (phase == 1) {
+    result = (struct wrapper *) c_zalloc(sizeof(struct wrapper) + *sizep);
```
Comment (Alkorin) [inline app/modules/mdns.c]: Never freed ?

**Comment by Alkorin** [comment]
It tooks me more than 10 minutes to understand what you tried to do with your phase 0 / phase 1.

You could split this func in two parts with more comments, it'll add readability to your code.
setptr is not obvious too... You use only 3 times, it's maybe better to write it like this

```
...
tmpBuffer = &result->data;
...
// Get server name
char * server_name = luaL_checkstring(L, 2)
// Store it
c_strcpy(tmpBuffer, server_name);
result->mdns_info.server_name = tmpBuffer;
tmpBuffer += c_strlen(server_name) + 1;
```
than
```
setptr(&result->mdns_info.server_name, &p, luaL_checkstring(L, 2));
```

It's easier to read.

**Comment by pjsg** [comment]
I cleaned up the code a bit and removed the setptr function. 

Also, I disabled the module by default. Turns out that the lwip code adds a fair amount of RAM consumption (maybe 700 bytes) if the mdns code is included.

**Comment by makefu** [comment]
i'd love to see this in mainline :+1:

**Comment by saeugetier** [comment]
I would like to see it in the mainline, too. It's a nice feature. In an DHCP environment you don't know each IP of each nodemcu module.

**Comment by pastukhov** [comment]
:+1:

**Comment by pastukhov** [comment]
Any progress?

**Comment by pastukhov** [comment]
Hi! @pjsg.
I have cherrypicked tis pr, but can't get this working.
I have add ```lua mdns.register('socket', "http", 80, { hardware='NodeMCU'})``` to my http server startup script and not able to find socket.local with avahi-discover or dig @224.0.0.251 -p 5353 socket.local

**Comment by pjsg** [comment]
I just built the branch again (against the current dev branch) and ran your mdns command line.

I then ran avahi-browse on my linux box:

avahi-browse --resolve _http._tcp
+ eth0 IPv4 socket                                        Web Site             local
+ eth0 IPv4 Living Room                                   Web Site             local
= eth0 IPv4 socket                                        Web Site             local
   hostname = [socket.local]
   address = [192.168.1.234]
   port = [80]
   txt = ["hardware=NodeMCU" "vendor = Espressif"]

You can see that 'socket' appears. Initially it didn't work until I modified my iptables configuration on my linux box to permit inbound UDP/5353 traffic. Note that the response doesn't come from the same address as the original packet destination, so the default return path doesn't work. Thus you need an explicit allow rule.

Also note that the service type is being prefixed by an underscore, so it doesn't matter if you include it in the registration or not.

Also, I think that you need to do the registration *after* you are connected to the network.

**Comment by FrankX0** [comment]
I am using mDNS with the latest dev build, but cannot get it stable to work.
The first few minutes the functionality works as expected, but after 5 to 10 minutes the service seems to stop. Through a web-browser I can no longer connect using the .local domain. Using dns-sd in Windows 10 also does not result in finding the IP address.
Any suggestions?

**Pull request merged**